pkg/app/server/apikeyverifier/verifier.go

```diff
@@ -23,24 +23,30 @@ import (
 
 	"github.com/pipe-cd/pipecd/pkg/cache"
 	"github.com/pipe-cd/pipecd/pkg/cache/memorycache"
+	"github.com/pipe-cd/pipecd/pkg/cache/rediscache"
 	"github.com/pipe-cd/pipecd/pkg/model"
+	"github.com/pipe-cd/pipecd/pkg/redis"
 )
 
 type apiKeyGetter interface {
 	Get(ctx context.Context, id string) (*model.APIKey, error)
 }
 
 type Verifier struct {
-	apiKeyCache cache.Cache
-	apiKeyStore apiKeyGetter
-	logger      *zap.Logger
+	apiKeyCache         cache.Cache
+	apiKeyStore         apiKeyGetter
+	apiKeyLastUsedCache cache.Cache
+	logger              *zap.Logger
 }
 
-func NewVerifier(ctx context.Context, getter apiKeyGetter, logger *zap.Logger) *Verifier {
+const apiKeyLastUsedCacheHashKey = "HASHKEY:PIPED:API_KEYS" //nolint:gosec
```

Comment:
[nits] I feel it's better to make it sharding per project.
Because it's supposed to be called per a project.

Reply:
Server must decide hash key name before NewVerifier is called, but which project is not decided at that time.
Piped client is associated with one project, but defaultPipedStatHashKey does not include project id because of the same reason.